File: android/src/main/java/com/reactnativebitchat/DeliveryTracker.java
package com.reactnativebitchat;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class DeliveryTracker {
    private final Map<String, String> messageStatuses = new HashMap<>();
    private final Map<String, List<DeliveryAck>> messageAcks = new HashMap<>();

    public void trackMessage(String messageID, String status) {
        if (status.equals("PENDING") || status.equals("DELIVERED") || status.equals("READ")) {
            messageStatuses.put(messageID, status);
        }
    }

    public DeliveryAck generateAck(String messageID, String recipientID, String nickname, int hopCount) {
        DeliveryAck ack = new DeliveryAck();
        ack.messageID = messageID;
        ack.recipientID = recipientID;
        ack.nickname = nickname;
        ack.hopCount = hopCount;
        ack.timestamp = System.currentTimeMillis();
        ack.id = UUID.randomUUID().toString();
        return ack;
    }

    public void processAck(DeliveryAck ack) {
        if (messageStatuses.containsKey(ack.messageID)) {
            if (ack.hopCount == 0) {
                trackMessage(ack.messageID, "DELIVERED");
            }
            messageAcks.computeIfAbsent(ack.messageID, k -> new ArrayList<>()).add(ack);
        }
    }

    public String getStatus(String messageID) {
        return messageStatuses.getOrDefault(messageID, "PENDING");
    }

    public List<DeliveryAck> getAcks(String messageID) {
        return new ArrayList<>(messageAcks.getOrDefault(messageID, new ArrayList<>()));
    }

    public static class DeliveryAck {
        public String id;
        public String messageID;
        public String recipientID;
        public String nickname;
        public int hopCount;
        public long timestamp;
    }
}

File: ios/DeliveryTracker.swift
import Foundation

class DeliveryTracker {
    private var messageStatuses = [String: String]()
    private var messageAcks = [String: [DeliveryAck]]()

    func trackMessage(_ messageID: String, status: String) {
        if ["PENDING", "DELIVERED", "READ"].contains(status) {
            messageStatuses[messageID] = status
        }
    }

    func generateAck(messageID: String, recipientID: String, nickname: String, hopCount: Int) -> DeliveryAck {
        var ack = DeliveryAck()
        ack.id = UUID().uuidString
        ack.messageID = messageID
        ack.recipientID = recipientID
        ack.nickname = nickname
        ack.hopCount = hopCount
        ack.timestamp = Date().timeIntervalSince1970 * 1000
        return ack
    }

    func processAck(_ ack: DeliveryAck) {
        if messageStatuses[ack.messageID] != nil {
            if ack.hopCount == 0 {
                trackMessage(ack.messageID, status: "DELIVERED")
            }
            var acks = messageAcks[ack.messageID, default: []]
            acks.append(ack)
            messageAcks[ack.messageID] = acks
        }
    }

    func getStatus(_ messageID: String) -> String? {
        return messageStatuses[messageID] ?? "PENDING"
    }

    func getAcks(_ messageID: String) -> [DeliveryAck] {
        return messageAcks[messageID] ?? []
    }

    struct DeliveryAck {
        var id: String
        var messageID: String
        var recipientID: String
        var nickname: String
        var hopCount: Int
        var timestamp: Double
    }
}